src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+    /// <returns>A bool on whether we sucessfully found a method declaration from <paramref name="invocation"/>.</returns>
+    private static bool TryGetMethodDeclaration(InvocationExpressionSyntax invocation, SyntaxNodeAnalysisContext context, out MethodDeclarationSyntax? methodDeclaration)
+    {
+        if (context.SemanticModel.GetSymbolInfo(invocation, context.CancellationToken).Symbol is not IMethodSymbol methodSymbol
+            || methodSymbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax(context.CancellationToken) is not MethodDeclarationSyntax methodDeclarationFound)
+        {
+            methodDeclaration = null;
+            return false;
+        }
+
+        methodDeclaration = methodDeclarationFound;
+        return true;
+    }
+
+    /// <summary>
+    /// Pushes all expression statements in a method declaration to a stack.
+    /// </summary>
+    /// <param name="methodDeclaration">The <see cref="MethodDeclarationSyntax"/> to find <see cref="ExpressionStatementSyntax"/> descendants.</param>
+    /// <param name="nodeStack">The <see cref="Stack{T}"/> to push <see cref="ExpressionStatementSyntax"/> nodes.</param>
+    private static void PushExpressionStatements(MethodDeclarationSyntax methodDeclaration, Stack<ExpressionStatementSyntax> nodeStack)
+    {
+        foreach (ExpressionStatementSyntax expressionStatement in methodDeclaration.DescendantNodes().OfType<ExpressionStatementSyntax>())
+        {
+            nodeStack.Push(expressionStatement);
+        }
+    }
+
+    /// <summary>
+    /// Processes an expression statement from a method declaration.
+    /// If the expression is an invocation, it retrieves the method declaration and pushes its expression statements onto the stack.
+    /// If the expression is an assignment to a field, it marks the field as initialized to avoid initialization in the declaration.
+    /// It ignores all other expressions.
+    /// </summary>
+    /// <param name="node">The expression statement to process.</param>
+    /// <param name="context">The context for the syntax node analysis.</param>
+    /// <param name="fields">A dictionary of field initialization information.</param>
+    /// <param name="nodeStack">A stack to hold expression statements for further processing.</param>
+    private static void ProcessExpressionStatementFromMethod(ExpressionStatementSyntax node, SyntaxNodeAnalysisContext context, Dictionary<string, FieldInitializationInfo> fields, Stack<ExpressionStatementSyntax> nodeStack)
+    {
+        if (node.Expression is InvocationExpressionSyntax invocationExpressionSyntax)
+        {
+            if (TryGetMethodDeclaration(invocationExpressionSyntax, context, out MethodDeclarationSyntax? nestedMethodDeclaration))
+            {
+                // We found an invocation expression to a method declaration, so we push the expression statements in the method to the stack.
+                PushExpressionStatements(nestedMethodDeclaration!, nodeStack);
```

Comment:
Don't use `!`, you could have null dereference. Do the nullability checks and bail if you don't have the proper types needed